cpp/biconnected_components_module/biconnected_components_test.cpp

```diff
@@ -0,0 +1,283 @@
+#include <chrono>
+#include <random>
+#include <set>
+#include <vector>
+
+#include <gtest/gtest.h>
+#include <mg_graph.hpp>
+
+#include "algorithm/biconnected_components.hpp"
+
+/// This class is threadsafe
+class Timer {
+ public:
+  Timer() : start_time_(std::chrono::steady_clock::now()) {}
+
+  template <typename TDuration = std::chrono::duration<double>>
+  TDuration Elapsed() const {
+    return std::chrono::duration_cast<TDuration>(std::chrono::steady_clock::now() - start_time_);
+  }
+
+ private:
+  std::chrono::steady_clock::time_point start_time_;
+};
+
+/// Builds the graph from a given number of nodes and a list of edges.
+/// Nodes should be 0-indexed and each edge should be provided in both
+/// directions.
+inline std::unique_ptr<mg_graph::Graph<>> BuildGraph(std::uint64_t nodes,
+                                                     std::vector<std::pair<std::uint64_t, std::uint64_t>> edges) {
+  auto G = std::make_unique<mg_graph::Graph<>>();
+  for (std::uint64_t i = 0; i < nodes; ++i) G->CreateNode(i);
+  for (auto &p : edges) G->CreateEdge(p.first, p.second);
+
+  return G;
+}
+
+/// Generates random undirected graph with a given numer of nodes and edges.
+/// The generated graph is not picked out of a uniform distribution.
+inline std::unique_ptr<mg_graph::Graph<>> GenRandomGraph(std::uint64_t nodes, std::uint64_t edges) {
+  auto seed = std::chrono::high_resolution_clock::now().time_since_epoch().count();
+  std::mt19937 rng(seed);
+  std::uniform_int_distribution<std::uint64_t> dist(0, nodes - 1);
+  std::set<std::pair<std::uint64_t, std::uint64_t>> E;
+  for (std::uint64_t i = 0; i < edges; ++i) {
+    std::int64_t u, v;
+    do {
+      u = dist(rng);
+      v = dist(rng);
+      if (u > v) std::swap(u, v);
+    } while (u == v || E.find({u, v}) != E.end());
+    E.insert({u, v});
```

Comment:
```suggestion
  for (std::uint64_t i = 0; i < edges; ++i) {
    std::optional<std::pair<std::uint64_t, std::uint64_t>> edge;
    do {
      edge = std::minmax(dist(rng), dist(rng));
    } while (edge->first == edge->second || E.find(*edge) != E.end());
    E.insert(*edge);
```

Reply:
You can also make an alias for the std::pair

Reply:
Where does this alias go in common practice? Inside of the header or implementation file?

Reply:
you can have it locally in this function for example, if you add it globally you could have naming conflicts and a lot of compiler troubles.